### PR TITLE
add string_id and simplify multi-handle container

### PR DESCRIPTION
**Pull request by havardpe**
add more testing
static instance instead of lazy singleton
use logging for leak detection
drop label_t type alias

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

@arnej27959 please review

### CODE REVIEW

File: vespalib/src/vespa/vespalib/util/string_id.h [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+/**
+ * A typed integer value representing the identity of a string stored
+ * in the global SharedStringRepo. This class is a simple wrapper with
+ * no lifetime management of the mapping between string value and
+ * string id. For a string_id to be valid, it needs to be owned by at
+ * least one SharedStringRepo::Handle or SharedStringRepo::Handles
+ * object. This is similar to how std::enable_shared_from_this works;
+ * the string_id acts like a reference to a mapping from a string to a
+ * numberical value (without ownership) and Handle/Handles act like
```
Comment (arnej27959) [inline vespalib/src/vespa/vespalib/util/string_id.h:19]: numberical -> numerical

File: vespalib/src/vespa/vespalib/util/shared_string_repo.h [hunk 2]
```diff
@@ -156,125 +165,107 @@ class SharedStringRepo {
...
-        Handle(vespalib::stringref str) : _id(get().resolve(str)) {}
-        Handle(const Handle &rhs) : _id(get().copy(rhs._id)) {}
+        Handle() noexcept : _id() {}
+        Handle(vespalib::stringref str) : _id(_repo.resolve(str)) {}
+        Handle(const Handle &rhs) : _id(_repo.copy(rhs._id)) {}
         Handle &operator=(const Handle &rhs) {
-            get().reclaim(_id);
-            _id = get().copy(rhs._id);
-            return *this;            
+            _repo.reclaim(_id);
```
Comment (arnej27959) [inline vespalib/src/vespa/vespalib/util/shared_string_repo.h:222]: do we need to worry about self-assignment? if so we would need to do
```
string_id copy = _repo.copy(rhs._id);
_repo.reclaim(_id);
_id = copy;
```

**Review by arnej27959 — APPROVED** [review]

**Comment by havardpe** [comment]
@arnej27959 PTAL

**Review by arnej27959 — APPROVED** [review]
👍

**Pull request merged**